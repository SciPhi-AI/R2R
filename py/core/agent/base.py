import asyncio
from abc import ABCMeta
from typing import AsyncGenerator, Generator, Optional

from core.base.abstractions import (
    AsyncSyncMeta,
    LLMChatCompletion,
    LLMChatCompletionChunk,
    Message,
    syncable,
)
from core.base.agent import Agent, Conversation
import logging

logger = logging.getLogger(__name__)


class CombinedMeta(AsyncSyncMeta, ABCMeta):
    pass


def sync_wrapper(async_gen):
    loop = asyncio.get_event_loop()

    def wrapper():
        try:
            while True:
                try:
                    yield loop.run_until_complete(async_gen.__anext__())
                except StopAsyncIteration:
                    break
        finally:
            loop.run_until_complete(async_gen.aclose())

    return wrapper()


class R2RAgent(Agent, metaclass=CombinedMeta):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._register_tools()
        self._reset()

    def _reset(self):
        self._completed = False
        self.conversation = Conversation()

    @syncable
    async def arun(
        self,
        system_instruction: Optional[str] = None,
        messages: Optional[list[Message]] = None,
        *args,
        **kwargs,
    ) -> list[dict]:
        self._reset()
        await self._setup(system_instruction)

        if messages:
            for message in messages:
                await self.conversation.add_message(message)

        while not self._completed:
            messages_list = await self.conversation.get_messages()
            generation_config = self.get_generation_config(messages_list[-1])
            response = await self.llm_provider.aget_completion(
                messages_list,
                generation_config,
            )
            await self.process_llm_response(response, *args, **kwargs)

        return await self.conversation.get_messages()

    async def process_llm_response(
        self, response: LLMChatCompletion, *args, **kwargs
    ) -> None:
        if not self._completed:
            message = response.choices[0].message
            if message.function_call:
                await self.handle_function_or_tool_call(
                    message.function_call.name,
                    message.function_call.arguments,
                    *args,
                    **kwargs,
                )
            elif message.tool_calls:
                for tool_call in message.tool_calls:
                    await self.handle_function_or_tool_call(
                        tool_call.function.name,
                        tool_call.function.arguments,
                        *args,
                        **kwargs,
                    )
            else:
                await self.conversation.add_message(
                    Message(role="assistant", content=message.content)
                )
                self._completed = True


class R2RStreamingAgent(R2RAgent):
    async def arun(  # type: ignore
        self,
        system_instruction: Optional[str] = None,
        messages: Optional[list[Message]] = None,
        *args,
        **kwargs,
    ) -> AsyncGenerator[str, None]:
        self._reset()
        await self._setup(system_instruction)

        if messages:
            for message in messages:
                await self.conversation.add_message(message)

        while not self._completed:
            messages_list = await self.conversation.get_messages()

            generation_config = self.get_generation_config(
                messages_list[-1], stream=True
            )
            stream = self.llm_provider.get_completion_stream(
                messages_list,
                generation_config,
            )
            async for chunk in self.process_llm_response(
                stream, *args, **kwargs
            ):
                yield chunk

    def run(
        self, system_instruction, messages, *args, **kwargs
    ) -> Generator[str, None, None]:
        return sync_wrapper(
            self.arun(system_instruction, messages, *args, **kwargs)
        )

    async def process_llm_response(  # type: ignore
        self,
        stream: Generator[LLMChatCompletionChunk, None, None],
        *args,
        **kwargs,
    ) -> AsyncGenerator[str, None]:
        function_name = None
        function_arguments = ""
        content_buffer = ""

        for chunk in stream:
            delta = chunk.choices[0].delta
            if delta.tool_calls:
                for tool_call in delta.tool_calls:
                    if not tool_call.function:
                        logger.info("Tool function not found in tool call.")
                        continue
                    name = tool_call.function.name
                    if not name:
                        logger.info("Tool name not found in tool call.")
                        continue
                    arguments = tool_call.function.arguments
                    if not arguments:
                        logger.info("Tool arguments not found in tool call.")
                        continue

                    results = await self.handle_function_or_tool_call(
                        name,
                        arguments,
                        # FIXME: tool_call.id,
                        *args,
                        **kwargs,
                    )

                    yield "<tool_call>"
                    yield f"<name>{name}</name>"
                    yield f"<arguments>{arguments}</arguments>"
                    yield f"<results>{results.llm_formatted_result}</results>"
                    yield "</tool_call>"

            if delta.function_call:
                if delta.function_call.name:
                    function_name = delta.function_call.name
                if delta.function_call.arguments:
                    function_arguments += delta.function_call.arguments
            elif delta.content:
                if content_buffer == "":
                    yield "<completion>"
                content_buffer += delta.content
                yield delta.content

            if chunk.choices[0].finish_reason == "function_call":
                if not function_name:
                    logger.info("Function name not found in function call.")
                    continue

                yield "<function_call>"
                yield f"<name>{function_name}</name>"
                yield f"<arguments>{function_arguments}</arguments>"
                tool_result = await self.handle_function_or_tool_call(
                    function_name, function_arguments, *args, **kwargs
                )
                if tool_result.stream_result:
                    yield f"<results>{tool_result.stream_result}</results>"
                else:
                    yield f"<results>{tool_result.llm_formatted_result}</results>"

                yield "</function_call>"

                function_name = None
                function_arguments = ""

            elif chunk.choices[0].finish_reason == "stop":
                if content_buffer:
                    await self.conversation.add_message(
                        Message(role="assistant", content=content_buffer)
                    )
                self._completed = True
                yield "</completion>"

        # Handle any remaining content after the stream ends
        if content_buffer and not self._completed:
            await self.conversation.add_message(
                Message(role="assistant", content=content_buffer)
            )
            self._completed = True
            yield "</completion>"
