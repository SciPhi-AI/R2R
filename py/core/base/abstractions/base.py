import asyncio


class AsyncSyncMeta(type):
    _event_loop = None  # Class-level shared event loop

    @classmethod
    def get_event_loop(cls):
        if cls._event_loop is None or cls._event_loop.is_closed():
            cls._event_loop = asyncio.new_event_loop()
            asyncio.set_event_loop(cls._event_loop)
        return cls._event_loop

    def __new__(cls, name, bases, dct):
        new_cls = super().__new__(cls, name, bases, dct)
        for attr_name, attr_value in dct.items():
            if asyncio.iscoroutinefunction(attr_value) and getattr(
                attr_value, "_syncable", False
            ):
                sync_method_name = attr_name[
                    1:
                ]  # Remove leading 'a' for sync method
                async_method = attr_value

                def make_sync_method(async_method):
                    def sync_wrapper(self, *args, **kwargs):
                        loop = cls.get_event_loop()
                        if not loop.is_running():
                            # Setup to run the loop in a background thread if necessary
                            # to prevent blocking the main thread in a synchronous call environment
                            from threading import Thread

                            result = None
                            exception = None

                            def run():
                                nonlocal result, exception
                                try:
                                    asyncio.set_event_loop(loop)
                                    result = loop.run_until_complete(
                                        async_method(self, *args, **kwargs)
                                    )
                                except Exception as e:
                                    exception = e
                                finally:
                                    generation_config = kwargs.get(
                                        "rag_generation_config", None
                                    )
                                    if (
                                        not generation_config
                                        or not generation_config.stream
                                    ):
                                        loop.run_until_complete(
                                            loop.shutdown_asyncgens()
                                        )
                                        loop.close()

                            thread = Thread(target=run)
                            thread.start()
                            thread.join()
                            if exception:
                                raise exception
                            return result
                        else:
                            # If there's already a running loop, schedule and execute the coroutine
                            future = asyncio.run_coroutine_threadsafe(
                                async_method(self, *args, **kwargs), loop
                            )
                            return future.result()

                    return sync_wrapper

                setattr(
                    new_cls, sync_method_name, make_sync_method(async_method)
                )
        return new_cls


def syncable(func):
    """Decorator to mark methods for synchronous wrapper creation."""
    func._syncable = True
    return func
