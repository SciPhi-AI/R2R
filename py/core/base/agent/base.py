from typing import Any, Callable, Dict, Optional

from ..abstractions import R2RSerializable


class Tool(R2RSerializable):
    name: str
    description: str
    results_function: Callable
    llm_format_function: Callable
    stream_function: Optional[Callable] = None
    parameters: Optional[Dict[str, Any]] = None

    class Config:
        arbitrary_types_allowed = True


class ToolResult(R2RSerializable):
    raw_result: Any
    llm_formatted_result: str
    stream_result: Optional[str] = None
