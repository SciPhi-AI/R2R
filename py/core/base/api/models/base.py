from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class ResultsWrapper(BaseModel, Generic[T]):
    results: T
