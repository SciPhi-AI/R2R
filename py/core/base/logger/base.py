import logging
from abc import abstractmethod
from datetime import datetime
from enum import Enum
from typing import Any, Optional, Tuple, Union
from uuid import UUID

from pydantic import BaseModel

from core.base import Message

from ..providers.base import Provider, ProviderConfig

logger = logging.getLogger()


class RunInfoLog(BaseModel):
    run_id: UUID
    run_type: str
    timestamp: datetime
    user_id: UUID


class PersistentLoggingConfig(ProviderConfig):
    provider: str = "local"
    log_table: str = "logs"
    log_info_table: str = "log_info"
    logging_path: Optional[str] = None

    def validate_config(self) -> None:
        pass

    @property
    def supported_providers(self) -> list[str]:
        return ["local", "postgres"]


class RunType(str, Enum):
    """Enumeration of the different types of runs."""

    RETRIEVAL = "RETRIEVAL"
    MANAGEMENT = "MANAGEMENT"
    INGESTION = "INGESTION"
    AUTH = "AUTH"
    UNSPECIFIED = "UNSPECIFIED"
    KG = "KG"


class PersistentLoggingProvider(Provider):
    @abstractmethod
    async def close(self):
        pass

    @abstractmethod
    async def log(
        self,
        run_id: UUID,
        key: str,
        value: str,
    ):
        pass

    @abstractmethod
    async def get_logs(
        self,
        run_ids: list[UUID],
        limit_per_run: int,
    ) -> list:
        pass

    @abstractmethod
    async def info_log(
        self,
        run_id: UUID,
        run_type: RunType,
        user_id: UUID,
    ):
        pass

    @abstractmethod
    async def get_info_logs(
        self,
        offset: int = 0,
        limit: int = 100,
        run_type_filter: Optional[RunType] = None,
        user_ids: Optional[list[UUID]] = None,
    ) -> list[RunInfoLog]:
        pass
