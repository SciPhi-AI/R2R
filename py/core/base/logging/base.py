from enum import Enum


class RunType(str, Enum):
    """Enumeration of the different types of runs."""

    RETRIEVAL = "RETRIEVAL"
    MANAGEMENT = "MANAGEMENT"
    INGESTION = "INGESTION"
    AUTH = "AUTH"
    UNSPECIFIED = "UNSPECIFIED"
    KG = "KG"
