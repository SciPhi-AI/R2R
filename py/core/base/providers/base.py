from abc import ABC, abstractmethod
from typing import Any, Optional, Type

from pydantic import BaseModel

from ..abstractions import R2RSerializable


class AppConfig(R2RSerializable):
    project_name: Optional[str] = None

    @classmethod
    def create(cls, *args, **kwargs):
        project_name = kwargs.get("project_name")
        return AppConfig(project_name=project_name)


class ProviderConfig(BaseModel, ABC):
    """A base provider configuration class"""

    app: AppConfig  # Add an app_config field
    extra_fields: dict[str, Any] = {}
    provider: Optional[str] = None

    class Config:
        arbitrary_types_allowed = True
        ignore_extra = True

    @abstractmethod
    def validate_config(self) -> None:
        pass

    @classmethod
    def create(cls: Type["ProviderConfig"], **kwargs: Any) -> "ProviderConfig":
        base_args = cls.model_fields.keys()
        filtered_kwargs = {
            k: v if v != "None" else None
            for k, v in kwargs.items()
            if k in base_args
        }
        instance = cls(**filtered_kwargs)
        for k, v in kwargs.items():
            if k not in base_args:
                instance.extra_fields[k] = v
        return instance

    @property
    @abstractmethod
    def supported_providers(self) -> list[str]:
        """Define a list of supported providers."""
        pass

    @classmethod
    def from_dict(
        cls: Type["ProviderConfig"], data: dict[str, Any]
    ) -> "ProviderConfig":
        """Create a new instance of the config from a dictionary."""
        return cls.create(**data)


class Provider(ABC):
    """A base provider class to provide a common interface for all providers."""

    def __init__(self, config: ProviderConfig, *args, **kwargs):
        if config:
            config.validate_config()
        self.config = config
