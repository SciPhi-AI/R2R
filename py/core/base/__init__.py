from .abstractions import *
from .agent import *
from .api.models import *
from .logging import *
from .parsers import *
from .pipeline import *
from .pipes import *
from .providers import *
from .utils import *

__all__ = [
    ## ABSTRACTIONS
    # Base abstractions
    "AsyncSyncMeta",
    "syncable",
    # Completion abstractions
    "CompletionRecord",
    "MessageType",
    # Document abstractions
    "DataType",
    "Document",
    "DocumentExtraction",
    "DocumentFragment",
    "DocumentInfo",
    "IngestionStatus",
    "RestructureStatus",
    "DocumentType",
    # Embedding abstractions
    "EmbeddingPurpose",
    "default_embedding_prefixes",
    # Exception abstractions
    "R2RDocumentProcessingError",
    "R2RException",
    # KG abstractions
    "Entity",
    "KGExtraction",
    "Triple",
    # LLM abstractions
    "GenerationConfig",
    "LLMChatCompletion",
    "LLMChatCompletionChunk",
    "RAGCompletion",
    # Prompt abstractions
    "Prompt",
    # Search abstractions
    "AggregateSearchResult",
    "KGLocalSearchResult",
    "KGGlobalSearchResult",
    "KGSearchResult",
    "KGSearchSettings",
    "VectorSearchResult",
    "VectorSearchSettings",
    # Restructure abstractions
    "KGCreationSettings",
    "KGEnrichmentSettings",
    # User abstractions
    "Token",
    "TokenData",
    "UserStats",
    # Vector abstractions
    "Vector",
    "VectorEntry",
    "VectorType",
    "StorageResult",
    ## AGENT
    # Agent abstractions
    "Agent",
    "AgentConfig",
    "Conversation",
    "Message",
    "Tool",
    "ToolResult",
    ## API
    # Auth Responses
    "GenericMessageResponse",
    "TokenResponse",
    "UserResponse",
    ## LOGGING
    # Basic types
    "RunType",
    "AnalysisTypes",
    "LogAnalytics",
    "LogAnalyticsConfig",
    "LogFilterCriteria",
    "LogProcessor",
    # Logging Providers
    "LocalRunLoggingProvider",
    "LoggingConfig",
    "PostgresLoggingConfig",
    "PostgresRunLoggingProvider",
    "RedisLoggingConfig",
    "RedisRunLoggingProvider",
    "RunLoggingSingleton",
    # Run Manager
    "RunManager",
    "manage_run",
    ## PARSERS
    # Base parser
    "AsyncParser",
    ## PIPELINE
    # Base pipeline
    "AsyncPipeline",
    ## PIPES
    "AsyncPipe",
    "AsyncState",
    "PipeType",
    ## PROVIDERS
    # Base provider classes
    "Provider",
    "ProviderConfig",
    # Auth provider
    "AuthConfig",
    "AuthProvider",
    # Chunking provider
    "R2RChunkingConfig",
    "UnstructuredChunkingConfig",
    "ChunkingConfig",
    "ChunkingProvider",
    "Strategy",
    # Crypto provider
    "CryptoConfig",
    "CryptoProvider",
    # Database providers
    "DatabaseConfig",
    "DatabaseProvider",
    "RelationalDBProvider",
    "VectorDBProvider",
    # Embedding provider
    "EmbeddingConfig",
    "EmbeddingProvider",
    # File provider
    "FileConfig",
    "FileProvider",
    # Knowledge Graph provider
    "KGConfig",
    "KGProvider",
    # LLM provider
    "CompletionConfig",
    "CompletionProvider",
    # Parsing provider
    "ParsingConfig",
    "ParsingProvider",
    "OverrideParser",
    # Prompt provider
    "PromptConfig",
    "PromptProvider",
    ## UTILS
    "RecursiveCharacterTextSplitter",
    "TextSplitter",
    "run_pipeline",
    "to_async_generator",
    "generate_run_id",
    "generate_id_from_label",
    "increment_version",
    "EntityType",
    "RelationshipType",
    "format_entity_types",
    "format_relations",
]
