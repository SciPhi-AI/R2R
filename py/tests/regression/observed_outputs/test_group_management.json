{
  "create_groups": {
    "error": "'R2RClient' object has no attribute 'create_group'"
  },
  "add_users_to_groups": {
    "error": "'R2RClient' object has no attribute 'add_user_to_group'"
  },
  "group_based_document_access": {
    "error": "'TestGroupManagement' object has no attribute 'group_id_1'"
  },
  "admin_ingest_documents": {
    "error": "'TestGroupManagement' object has no attribute 'group_id_1'"
  },
  "user_ingest_and_search": {
    "error": "'TestGroupManagement' object has no attribute 'group_id_1'"
  },
  "cleanup": {
    "error": "'R2RClient' object has no attribute 'delete_group'"
  }
}
