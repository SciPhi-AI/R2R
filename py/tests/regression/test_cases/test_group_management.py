import uuid

from tests.regression.test_cases.base import BaseTest


class TestGroupManagement(BaseTest):
    def __init__(self, client):
        super().__init__(client)
        self.admin_email = "admin@example.com"
        self.admin_password = "change_me_immediately"
        self.user_password = "test_password"

        keys_to_ignore = ["group_id", "created_at", "updated_at"]
        self.set_exclude_paths(
            "create_groups_test",
            [f"root['group_1']['results']['{key}']" for key in keys_to_ignore]
            + [
                f"root['group_1']['results']['{key}']"
                for key in keys_to_ignore
            ],
        )

        keys_to_ignore += ["hashed_password"]
        self.set_exclude_paths(
            "add_users_to_groups_test",
            [f"root['user_1']['results']['{key}']" for key in keys_to_ignore]
            + [
                f"root['group_1']['results']['{key}']"
                for key in keys_to_ignore
            ],
        )

    def get_test_cases(self):
        return {
            "create_groups": self.create_groups_test,
            "add_users_to_groups": self.add_users_to_groups_test,
            "group_based_document_access": self.group_based_document_access_test,
            "admin_ingest_documents": self.admin_ingest_documents_test,
            "user_ingest_and_search": self.user_ingest_and_search_test,
            "cleanup": self.cleanup_test,
        }

    def create_groups_test(self, client):
        try:
            client.login(self.admin_email, self.admin_password)
            group_name_1 = f"Test Group 1 {uuid.uuid4()}"
            group_name_2 = f"Test Group 2 {uuid.uuid4()}"
            group_1 = client.create_group(
                group_name_1, "A test group for permissions"
            )
            group_2 = client.create_group(
                group_name_2, "Another test group for permissions"
            )
            self.group_id_1 = group_1["results"]["group_id"]
            self.group_id_2 = group_2["results"]["group_id"]
            return {"group_1": group_1, "group_2": group_2}
        except Exception as e:
            return {"error": str(e)}

    def add_users_to_groups_test(self, client):
        try:
            self.user_email_1 = f"user1_{uuid.uuid4()}@example.com"
            self.user_email_2 = f"user2_{uuid.uuid4()}@example.com"
            user_1 = client.register(self.user_email_1, self.user_password)
            user_2 = client.register(self.user_email_2, self.user_password)
            self.user_id_1 = user_1["results"]["id"]
            self.user_id_2 = user_2["results"]["id"]

            client.add_user_to_group(self.user_id_1, self.group_id_1)
            client.add_user_to_group(self.user_id_2, self.group_id_2)
            client.add_user_to_group(self.user_id_2, self.group_id_1)

            return {"user_1": user_1, "user_2": user_2}
        except Exception as e:
            return {"error": str(e)}

    def group_based_document_access_test(self, client):
        try:
            # Admin ingests a document for group 1
            client.login(self.admin_email, self.admin_password)
            admin_ingest = client.ingest_files(
                file_paths=["tests/regression/test_data/test_document.txt"],
                metadatas=[{"group_ids": [str(self.group_id_1)]}],
            )

            # User 1 searches for documents
            client.login(self.user_email_1, self.user_password)
            user_1_search = client.search("test document")

            # User 2 searches for documents
            client.login(self.user_email_2, self.user_password)
            user_2_search = client.search("test document")

            return {
                "admin_ingest": admin_ingest,
                "user_1_search": user_1_search,
                "user_2_search": user_2_search,
            }
        except Exception as e:
            return {"error": str(e)}

    def admin_ingest_documents_test(self, client):
        try:
            client.login(self.admin_email, self.admin_password)

            # Admin ingests a document for group 1
            admin_ingest_group1 = client.ingest_files(
                file_paths=[
                    "tests/regression/test_data/admin_document_group1.txt"
                ],
                metadatas=[{"group_ids": [str(self.group_id_1)]}],
            )

            # Admin ingests a document for user 1
            admin_ingest_user1 = client.ingest_files(
                file_paths=[
                    "tests/regression/test_data/admin_document_user1.txt"
                ]
            )

            return {
                "admin_ingest_group1": admin_ingest_group1,
                "admin_ingest_user1": admin_ingest_user1,
            }
        except Exception as e:
            return {"error": str(e)}

    def user_ingest_and_search_test(self, client):
        try:
            # User 1 actions
            client.login(self.user_email_1, self.user_password)
            user_1_ingest = client.ingest_files(
                file_paths=["tests/regression/test_data/user1_document.txt"]
            )
            user_1_ingest_group = client.ingest_files(
                file_paths=[
                    "tests/regression/test_data/user1_document_group.txt"
                ],
                metadatas=[{"group_ids": [str(self.group_id_1)]}],
            )

            user_1_search = client.search("document")

            # User 2 actions
            client.login(self.user_email_2, self.user_password)
            user_2_ingest = client.ingest_files(
                file_paths=["tests/regression/test_data/user2_document.txt"]
            )
            user_2_search = client.search("document")

            return {
                "user_1_ingest": user_1_ingest,
                "user_1_search": user_1_search,
                "user_2_ingest": user_2_ingest,
                "user_2_search": user_2_search,
            }
        except Exception as e:
            return {"error": str(e)}

    def cleanup_test(self, client):
        try:
            client.login(self.admin_email, self.admin_password)
            client.delete_group(self.group_id_1)
            client.delete_group(self.group_id_2)
            client.delete_user(self.user_id_1, self.user_password)
            client.delete_user(self.user_id_2, self.user_password)
            return {"status": "cleanup completed"}
        except Exception as e:
            return {"error": str(e)}
