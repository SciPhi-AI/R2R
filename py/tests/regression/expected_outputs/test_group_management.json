{
  "create_groups": {
    "group_1": {
      "results": {
        "group_id": "3b24e22c-843d-4cd1-8183-6f244211542b",
        "name": "Test Group 1 90248f87-cf4f-42ac-9f10-c68716e46ccb",
        "description": "A test group for permissions",
        "created_at": "2024-08-14T01:01:19.662617+00:00",
        "updated_at": "2024-08-14T01:01:19.662617+00:00"
      }
    },
    "group_2": {
      "results": {
        "group_id": "3a2bfab3-5527-47a1-a901-a463e9b06971",
        "name": "Test Group 2 99212e13-64dc-456a-93d3-3e2afc97f3be",
        "description": "Another test group for permissions",
        "created_at": "2024-08-14T01:01:19.803314+00:00",
        "updated_at": "2024-08-14T01:01:19.803314+00:00"
      }
    }
  },
  "add_users_to_groups": {
    "user_1": {
      "results": {
        "email": "user1_62951675-cbc1-40dd-b005-3758adfc4c68@example.com",
        "id": "a34207e5-9b1a-5fe3-8a02-9ff56464b112",
        "group_ids": [],
        "hashed_password": "$2b$12$tVPpwO.Tr5KR/gtWm1UPDuq6.pXY3rlO7YonLYR0XcwwJkhgVI96O",
        "is_superuser": false,
        "is_active": true,
        "is_verified": false,
        "verification_code_expiry": null,
        "name": null,
        "bio": null,
        "profile_picture": null,
        "created_at": "2024-08-14T01:01:20.181505Z",
        "updated_at": "2024-08-14T01:01:20.181505Z"
      }
    },
    "user_2": {
      "results": {
        "email": "user2_16f40e51-174a-4b06-b308-1160b701cda7@example.com",
        "id": "4404314d-12a1-5299-9f7a-adfac07a5a3b",
        "group_ids": [],
        "hashed_password": "$2b$12$oqVl4mg7sL6UfUO2ziz4suSSOOBwYJisag.E.pHNWxd0YMde57e0y",
        "is_superuser": false,
        "is_active": true,
        "is_verified": false,
        "verification_code_expiry": null,
        "name": null,
        "bio": null,
        "profile_picture": null,
        "created_at": "2024-08-14T01:01:20.651635Z",
        "updated_at": "2024-08-14T01:01:20.651635Z"
      }
    }
  },
  "group_based_document_access": {
    "error": "An error ''PostgresVectorDBProvider' object has no attribute 'assign_document_to_group'' occurred during ingest_files_app"
  },
  "admin_ingest_documents": {
    "error": "An error ''PostgresVectorDBProvider' object has no attribute 'assign_document_to_group'' occurred during ingest_files_app"
  },
  "user_ingest_and_search": {
    "user_1_ingest": {
      "results": {
        "processed_documents": [
          {
            "id": "55d7b67e-c717-5e89-a956-61580475199d",
            "group_ids": [],
            "user_id": "a34207e5-9b1a-5fe3-8a02-9ff56464b112",
            "type": "txt",
            "metadata": {},
            "title": "user1_document.txt",
            "version": "v0",
            "size_in_bytes": 15,
            "status": "success",
            "created_at": "2024-08-13T18:01:23.617547",
            "updated_at": "2024-08-13T18:01:23.617547"
          }
        ],
        "failed_documents": [],
        "skipped_documents": []
      }
    },
    "user_1_search": {
      "results": {
        "vector_search_results": [
          {
            "fragment_id": "ced2d47c-524c-58d3-8cc3-0d474312fb00",
            "extraction_id": "b4451d80-760d-5e0f-93bc-cc0a89b1630d",
            "document_id": "55d7b67e-c717-5e89-a956-61580475199d",
            "user_id": "a34207e5-9b1a-5fe3-8a02-9ff56464b112",
            "group_ids": [],
            "score": 0.577595285558281,
            "text": "user1_document",
            "metadata": {
              "text": "user1_document",
              "title": "user1_document.txt",
              "associatedQuery": "document"
            }
          },
          {
            "fragment_id": "2b30d8c0-d037-5ca7-9961-08e2a13a25cd",
            "extraction_id": "eae05bf5-f732-53b3-80e5-a6e39d5a23d3",
            "document_id": "653c933c-867d-5588-b6cd-54d9412a8ffa",
            "user_id": "a34207e5-9b1a-5fe3-8a02-9ff56464b112",
            "group_ids": [],
            "score": 0.487762882211468,
            "text": "user1_document_group",
            "metadata": {
              "text": "user1_document_group",
              "title": "user1_document_group.txt",
              "user_id": "a34207e5-9b1a-5fe3-8a02-9ff56464b112",
              "associatedQuery": "document"
            }
          }
        ],
        "kg_search_results": []
      }
    },
    "user_2_ingest": {
      "results": {
        "processed_documents": [
          {
            "id": "6a7d57a8-0bab-55df-8674-a94b1ecd6492",
            "group_ids": [],
            "user_id": "4404314d-12a1-5299-9f7a-adfac07a5a3b",
            "type": "txt",
            "metadata": {},
            "title": "user2_document.txt",
            "version": "v0",
            "size_in_bytes": 15,
            "status": "success",
            "created_at": "2024-08-13T18:01:25.577549",
            "updated_at": "2024-08-13T18:01:25.577549"
          }
        ],
        "failed_documents": [],
        "skipped_documents": []
      }
    },
    "user_2_search": {
      "results": {
        "vector_search_results": [
          {
            "fragment_id": "4f3e93df-099c-58a7-a5cf-c40ba5ae76c1",
            "extraction_id": "838aa00a-2d5a-588c-9aa1-2553ae514024",
            "document_id": "6a7d57a8-0bab-55df-8674-a94b1ecd6492",
            "user_id": "4404314d-12a1-5299-9f7a-adfac07a5a3b",
            "group_ids": [],
            "score": 0.530904515656706,
            "text": "user2_document",
            "metadata": {
              "text": "user2_document",
              "title": "user2_document.txt",
              "associatedQuery": "document"
            }
          }
        ],
        "kg_search_results": []
      }
    }
  },
  "cleanup": {
    "status": "cleanup completed"
  }
}
